File: backend/app/services/generation/loading_title.py
"""Helpers for compact, readable loading-state generation titles."""

from __future__ import annotations

import os
import re
from typing import Iterable

DEFAULT_LOADING_TITLE = "生成中..."
_MAX_LOADING_TITLE_CHARS = 18
_LEADING_PROMPT_RE = re.compile(
    r"^(?:请(?:帮我)?|帮我|麻烦|需要|想要|我要|我想|请你|请为我|为我|给我)?"
    r"(?:基于以下内容|根据以下内容|围绕以下内容|结合以下内容)?"
    r"(?:设计|准备|创建|生成|制作|输出|整理|撰写|写|做)(?:一份|一个|一套|一页|一组|一版|份|个|套)?",
    re.IGNORECASE,
)
_TRAILING_SUFFIX_RE = re.compile(
    r"(?:的)?(?:演示文稿|演示稿|幻灯片|PPTX?|pptx?|汇报稿|汇报|分享稿|分享|报告|培训材料|方案)"
    r"(?:初稿|草稿|大纲|内容)?$",
    re.IGNORECASE,
)
_PAGE_HINT_RE = re.compile(r"(?:\d+\s*(?:页|slides?|pages?)\s*)+$", re.IGNORECASE)
_WHITESPACE_RE = re.compile(r"\s+")


def _trim_source_name(name: str) -> str:
    trimmed = str(name or "").strip()
    if not trimmed:
        return ""
    stem, ext = os.path.splitext(trimmed)
    candidate = stem or trimmed
    return candidate.strip()


def _strip_prefix_and_suffix(text: str) -> str:
    candidate = text.strip().strip("\"'“”‘’[]()【】")
    candidate = _LEADING_PROMPT_RE.sub("", candidate).strip(" ：:，,。；;、")
    candidate = _TRAILING_SUFFIX_RE.sub("", candidate).strip(" ：:，,。；;、")
    candidate = _PAGE_HINT_RE.sub("", candidate).strip(" ：:，,。；;、")
    candidate = re.sub(r"的+$", "", candidate).strip(" ：:，,。；;、")
    return candidate


def compact_loading_title(text: str | None, fallback: str = DEFAULT_LOADING_TITLE) -> str:
    normalized = _WHITESPACE_RE.sub(" ", str(text or "")).strip()
    if not normalized:
        return fallback

    first_line = next((part.strip() for part in normalized.splitlines() if part.strip()), "")
    candidate = first_line or normalized

    if "关于" in candidate:
        candidate = candidate.rsplit("关于", 1)[-1].strip()
    elif "围绕" in candidate:
        candidate = candidate.rsplit("围绕", 1)[-1].strip()
    elif "聚焦" in candidate:
        candidate = candidate.rsplit("聚焦", 1)[-1].strip()
    elif "主题" in candidate and re.search(r"主题\s*[:：是为]", candidate):
        candidate = re.split(r"主题\s*[:：是为]", candidate, maxsplit=1)[-1].strip()

    candidate = re.split(r"[。！？!?；;，,\n]", candidate, maxsplit=1)[0].strip()
    candidate = _strip_prefix_and_suffix(candidate)

    if not candidate:
        candidate = _strip_prefix_and_suffix(first_line or normalized)
    if not candidate:
        return fallback

    if len(candidate) <= _MAX_LOADING_TITLE_CHARS:
        return candidate
    return f"{candidate[:_MAX_LOADING_TITLE_CHARS].rstrip()}..."


def build_loading_title(
    *,
    topic: str | None = None,
    source_names: Iterable[str] | None = None,
    fallback: str = DEFAULT_LOADING_TITLE,
) -> str:
    topic_title = compact_loading_title(topic, fallback="")
    if topic_title:
        return topic_title

    cleaned_names = [_trim_source_name(name) for name in (source_names or [])]
    cleaned_names = [name for name in cleaned_names if name]
    if len(cleaned_names) == 1:
        return compact_loading_title(cleaned_names[0], fallback=fallback)
    if len(cleaned_names) > 1:
        return f"基于{len(cleaned_names)}个来源生成"
    return fallback

File: frontend/src/lib/loading-title.ts
const DEFAULT_LOADING_TITLE = "生成中...";
const MAX_LOADING_TITLE_CHARS = 18;

function stripSourceExtension(name: string): string {
  const trimmed = name.trim();
  if (!trimmed) return "";
  return trimmed.replace(/\.[^/.]+$/, "").trim();
}

function stripPromptBoilerplate(text: string): string {
  return text
    .trim()
    .replace(/^[\"'“”‘’【】\[\]()]+|[\"'“”‘’【】\[\]()]+$/g, "")
    .replace(
      /^(?:请(?:帮我)?|帮我|麻烦|需要|想要|我要|我想|请你|请为我|为我|给我)?(?:基于以下内容|根据以下内容|围绕以下内容|结合以下内容)?(?:设计|准备|创建|生成|制作|输出|整理|撰写|写|做)(?:一份|一个|一套|一页|一组|一版|份|个|套)?/i,
      ""
    )
    .replace(
      /(?:的)?(?:演示文稿|演示稿|幻灯片|PPTX?|pptx?|汇报稿|汇报|分享稿|分享|报告|培训材料|方案)(?:初稿|草稿|大纲|内容)?$/i,
      ""
    )
    .replace(/(?:\d+\s*(?:页|slides?|pages?)\s*)+$/i, "")
    .trim()
    .replace(/的+$/g, "")
    .trim()
    .replace(/^[：:，,。；;、]+|[：:，,。；;、]+$/g, "");
}

export function compactLoadingTitle(
  input: string | null | undefined,
  fallback = DEFAULT_LOADING_TITLE
): string {
  const normalized = String(input ?? "")
    .replace(/\s+/g, " ")
    .trim();
  if (!normalized) return fallback;

  const firstLine =
    normalized
      .split(/\r?\n/)
      .map((part) => part.trim())
      .find(Boolean) ?? "";

  let candidate = firstLine || normalized;
  if (candidate.includes("关于")) {
    candidate = candidate.split("关于").pop() ?? candidate;
  } else if (candidate.includes("围绕")) {
    candidate = candidate.split("围绕").pop() ?? candidate;
  } else if (candidate.includes("聚焦")) {
    candidate = candidate.split("聚焦").pop() ?? candidate;
  } else if (/主题\s*[:：是为]/.test(candidate)) {
    candidate = candidate.split(/主题\s*[:：是为]/).pop() ?? candidate;
  }

  candidate = candidate.split(/[。！？!?；;，,\n]/, 1)[0]?.trim() ?? candidate;
  candidate = stripPromptBoilerplate(candidate);
  if (!candidate) {
    candidate = stripPromptBoilerplate(firstLine || normalized);
  }
  if (!candidate) return fallback;

  if (candidate.length <= MAX_LOADING_TITLE_CHARS) {
    return candidate;
  }
  return `${candidate.slice(0, MAX_LOADING_TITLE_CHARS).trimEnd()}...`;
}

export function buildLoadingTitle(input: {
  topic?: string | null;
  sourceNames?: string[] | null;
  fallback?: string;
}): string {
  const fallback = input.fallback ?? DEFAULT_LOADING_TITLE;
  const topicTitle = compactLoadingTitle(input.topic, "");
  if (topicTitle) return topicTitle;

  const sourceNames = (input.sourceNames ?? [])
    .map((name) => stripSourceExtension(String(name ?? "")))
    .filter(Boolean);
  if (sourceNames.length === 1) {
    return compactLoadingTitle(sourceNames[0], fallback);
  }
  if (sourceNames.length > 1) {
    return `基于${sourceNames.length}个来源生成`;
  }
  return fallback;
}

export function resolveGenerationRequestTitle(
  request:
    | {
        title?: string | null;
        topic?: string | null;
      }
    | null
    | undefined,
  fallback = DEFAULT_LOADING_TITLE
): string {
  const topicTitle = compactLoadingTitle(request?.topic, "");
  if (topicTitle) return topicTitle;
  return compactLoadingTitle(request?.title, fallback);
}

export { DEFAULT_LOADING_TITLE };
